File: tests/omnia/data/structure/mutable/LinkedQueueTest.java
package omnia.data.structure.mutable;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

import java.util.Optional;
import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class LinkedQueueTest {

  @Test
  public void init_isEmpty() {
    Queue<?> testSubject = new LinkedQueue<>();

    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void init_countIsZero() {
    Queue<?> testSubject = new LinkedQueue<>();

    assertEquals(0, testSubject.count());
  }

  @Test
  public void enqueue_one_isPopulated() {
    Queue<Object> testSubject = new LinkedQueue<>();

    testSubject.enqueue(new Object());

    assertTrue(testSubject.isPopulated());
  }

  @Test
  public void enqueue_one_countIsOne() {
    Queue<Object> testSubject = new LinkedQueue<>();

    testSubject.enqueue(new Object());

    assertEquals(1, testSubject.count());
  }

  @Test(expected = NullPointerException.class)
  public void enqueue_null_didThrowException() {
    Queue<Object> testSubject = new LinkedQueue<>();

    testSubject.enqueue(null);
  }

  @Test
  public void enqueue_thenDequeue_didReturnObject() {
    Queue<Object> testSubject = new LinkedQueue<>();
    Object item = new Object();
    testSubject.enqueue(item);

    Optional<Object> result = testSubject.dequeue();

    assertTrue(result.isPresent());
    assertSame(item, result.get());
  }

  @Test
  public void enqueue_thenDequeue_isEmpty() {
    Queue<Object> testSubject = new LinkedQueue<>();
    Object item = new Object();
    testSubject.enqueue(item);

    testSubject.dequeue();

    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void enqueue_thenDequeue_countIsZero() {
    Queue<Object> testSubject = new LinkedQueue<>();
    Object item = new Object();
    testSubject.enqueue(item);

    testSubject.dequeue();

    assertEquals(0, testSubject.count());
  }

  @Test
  public void dequeue_whenEmpty_didReturnEmpty() {
    assertFalse(new LinkedQueue<>().dequeue().isPresent());
  }

  @Test
  public void enqueue_thenDequeueTwice_didReturnEmpty() {
    Queue<Object> testSubject = new LinkedQueue<>();
    testSubject.enqueue(new Object());

    testSubject.dequeue();

    assertFalse(testSubject.dequeue().isPresent());
  }

  @Test
  public void enqueue_thenDequeue_fiveHundredTimes_didReturnItems() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = new LinkedQueue<>();

    for (int i = 0; i < data.count(); i++) {
      testSubject.enqueue(i);

      assertEquals(data.itemAt(i), testSubject.dequeue().get());
    }
  }

  @Test
  public void enqueue_thenDequeue_fiveHundredTimes_isEmpty() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = new LinkedQueue<>();

    for (int i = 0; i < data.count(); i++) {
      testSubject.enqueue(i);
      testSubject.dequeue();
    }

    assertEquals(0, testSubject.count());
    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void enqueue_fiveHundredTimes_thenDequeue_fiveHundredTimes_didReturnItems() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = new LinkedQueue<>();

    for (int i = 0; i < data.count(); i++) {
      testSubject.enqueue(i);
    }

    for (int i = 0; i < data.count(); i++) {
      assertEquals(data.itemAt(i), testSubject.dequeue().get());
    }
  }

  @Test
  public void enqueue_fiveHundredTimes_thenDequeue_fiveHundredTimes_isEmpty() {
    List<Integer> data = buildData(500);
    Queue<Integer> testSubject = new LinkedQueue<>();

    for (int i = 0; i < data.count(); i++) {
      testSubject.enqueue(i);
    }
    for (int i = 0; i < data.count(); i++) {
      testSubject.dequeue();
    }

    assertEquals(0, testSubject.count());
    assertFalse(testSubject.isPopulated());
  }

  @Test
  public void peek_whenEmpty_isEmpty() {
    assertFalse(new LinkedQueue<>().peek().isPresent());
  }

  @Test
  public void peek_withOne_didReturnObject() {
    Queue<Integer> testSubject = new LinkedQueue<>();
    testSubject.enqueue(132);

    Optional<Integer> datum = testSubject.peek();

    assertTrue(datum.isPresent());
    assertEquals(Integer.valueOf(132), datum.get());
  }

  @Test
  public void peek_withOne_didNotDequeue() {
    Queue<Object> testSubject = new LinkedQueue<>();
    testSubject.enqueue(new Object());

    testSubject.peek();

    assertEquals(1, testSubject.count());
    assertTrue(testSubject.isPopulated());
  }

  @Test
  public void peekRepeatedly_withContent_didNotDequeue() {
    Queue<Object> testSubject = new LinkedQueue<>();
    for (int i = 0; i < 10; i++) {
      testSubject.enqueue(new Object());
    }

    for (int i = 0; i < 200; i++) {
      testSubject.peek();
    }

    assertEquals(10, testSubject.count());
    assertTrue(testSubject.isPopulated());
  }

  @Test
  public void peekRepeatedly_withContent_didReturnFirst() {
    Queue<Object> testSubject = new LinkedQueue<>();
    // First item (which we'll compare with)
    Object expected = new Object();
    testSubject.enqueue(expected);
    // Additional arbitrary objects
    for (int i = 0; i < 10; i++) {
      testSubject.enqueue(new Object());
    }

    for (int i = 0; i < 200; i++) {
      Optional<Object> result = testSubject.peek();
      assertTrue(result.isPresent());
      assertEquals(expected, result.get());
    }
  }

  private static List<Integer> buildData(int count) {
    if (count < 0) {
      throw new IllegalArgumentException(String.format("count must be positive (was %d)", count));
    }
    ImmutableList.Builder<Integer> data = ImmutableList.builder();
    for (int i = 0; i < count; i++) {
      data.add(i);
    }
    return data.build();
  }
}

File: tests/omnia/data/structure/mutable/LinkedQueueTest.kt
package omnia.data.structure.mutable

import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class LinkedQueueTest {

  @Test
  fun init_isEmpty() {
    val testSubject: Queue<*> = LinkedQueue<Any>()
    Assert.assertFalse(testSubject.isPopulated)
  }

  @Test
  fun init_countIsZero() {
    val testSubject: Queue<*> = LinkedQueue<Any>()
    Assert.assertEquals(0, testSubject.count().toLong())
  }

  @Test
  fun enqueue_one_isPopulated() {
    val testSubject: Queue<Any> = LinkedQueue()
    testSubject.enqueue(Any())
    Assert.assertTrue(testSubject.isPopulated)
  }

  @Test
  fun enqueue_one_countIsOne() {
    val testSubject: Queue<Any> = LinkedQueue()
    testSubject.enqueue(Any())
    Assert.assertEquals(1, testSubject.count().toLong())
  }

  @Test
  fun enqueue_thenDequeue_didReturnObject() {
    val testSubject: Queue<Any> = LinkedQueue()
    val item = Any()
    testSubject.enqueue(item)
    val result = testSubject.dequeue()
    Assert.assertTrue(result.isPresent)
    Assert.assertSame(item, result.get())
  }

  @Test
  fun enqueue_thenDequeue_isEmpty() {
    val testSubject: Queue<Any> = LinkedQueue()
    val item = Any()
    testSubject.enqueue(item)
    testSubject.dequeue()
    Assert.assertFalse(testSubject.isPopulated)
  }

  @Test
  fun enqueue_thenDequeue_countIsZero() {
    val testSubject: Queue<Any> = LinkedQueue()
    val item = Any()
    testSubject.enqueue(item)
    testSubject.dequeue()
    Assert.assertEquals(0, testSubject.count().toLong())
  }

  @Test
  fun dequeue_whenEmpty_didReturnEmpty() {
    Assert.assertFalse(LinkedQueue<Any>().dequeue().isPresent)
  }

  @Test
  fun enqueue_thenDequeueTwice_didReturnEmpty() {
    val testSubject: Queue<Any> = LinkedQueue()
    testSubject.enqueue(Any())
    testSubject.dequeue()
    Assert.assertFalse(testSubject.dequeue().isPresent)
  }

  @Test
  fun enqueue_thenDequeue_fiveHundredTimes_didReturnItems() {
    val data = buildData(500)
    val testSubject: Queue<Int> = LinkedQueue()
    for (i in 0 until data.count()) {
      testSubject.enqueue(i)
      Assert.assertEquals(data.itemAt(i), testSubject.dequeue().get())
    }
  }

  @Test
  fun enqueue_thenDequeue_fiveHundredTimes_isEmpty() {
    val data = buildData(500)
    val testSubject: Queue<Int> = LinkedQueue()
    for (i in 0 until data.count()) {
      testSubject.enqueue(i)
      testSubject.dequeue()
    }
    Assert.assertEquals(0, testSubject.count().toLong())
    Assert.assertFalse(testSubject.isPopulated)
  }

  @Test
  fun enqueue_fiveHundredTimes_thenDequeue_fiveHundredTimes_didReturnItems() {
    val data = buildData(500)
    val testSubject: Queue<Int> = LinkedQueue()
    for (i in 0 until data.count()) {
      testSubject.enqueue(i)
    }
    for (i in 0 until data.count()) {
      Assert.assertEquals(data.itemAt(i), testSubject.dequeue().get())
    }
  }

  @Test
  fun enqueue_fiveHundredTimes_thenDequeue_fiveHundredTimes_isEmpty() {
    val data = buildData(500)
    val testSubject: Queue<Int> = LinkedQueue()
    for (i in 0 until data.count()) {
      testSubject.enqueue(i)
    }
    for (i in 0 until data.count()) {
      testSubject.dequeue()
    }
    Assert.assertEquals(0, testSubject.count().toLong())
    Assert.assertFalse(testSubject.isPopulated)
  }

  @Test
  fun peek_whenEmpty_isEmpty() {
    Assert.assertFalse(LinkedQueue<Any>().peek().isPresent)
  }

  @Test
  fun peek_withOne_didReturnObject() {
    val testSubject: Queue<Int> = LinkedQueue()
    testSubject.enqueue(132)
    val datum = testSubject.peek()
    Assert.assertTrue(datum.isPresent)
    Assert.assertEquals(Integer.valueOf(132), datum.get())
  }

  @Test
  fun peek_withOne_didNotDequeue() {
    val testSubject: Queue<Any> = LinkedQueue()
    testSubject.enqueue(Any())
    testSubject.peek()
    Assert.assertEquals(1, testSubject.count().toLong())
    Assert.assertTrue(testSubject.isPopulated)
  }

  @Test
  fun peekRepeatedly_withContent_didNotDequeue() {
    val testSubject: Queue<Any> = LinkedQueue()
    for (i in 0..9) {
      testSubject.enqueue(Any())
    }
    for (i in 0..199) {
      testSubject.peek()
    }
    Assert.assertEquals(10, testSubject.count().toLong())
    Assert.assertTrue(testSubject.isPopulated)
  }

  @Test
  fun peekRepeatedly_withContent_didReturnFirst() {
    val testSubject: Queue<Any> = LinkedQueue()
    // First item (which we'll compare with)
    val expected = Any()
    testSubject.enqueue(expected)
    // Additional arbitrary objects
    for (i in 0..9) {
      testSubject.enqueue(Any())
    }
    for (i in 0..199) {
      val result = testSubject.peek()
      Assert.assertTrue(result.isPresent)
      Assert.assertEquals(expected, result.get())
    }
  }

  companion object {

    private fun buildData(count: Int): List<Int> {
      require(count >= 0) { String.format("count must be positive (was %d)", count) }
      val data = ImmutableList.builder<Int>()
      for (i in 0 until count) {
        data.add(i)
      }
      return data.build()
    }
  }
}